ENH: `scipy.stats` now returns namedtuples.

aeklant opened the pull request:
See issue #4192

aeklant commented:
Hello all,

This PR is still a work in progress. I aim to address all of the functions outlined by @clarkfitzg in issue #4192 but I wanted to get feedback from the community before I dive in and do it for all the functions (to avoid rework as much as possible). Basically all of the functions and tests will end up looking pretty much like this (with maybe a few differences in style to keep consistency within the module).

Here are a few notes:
1) For the functions that are in both `stats` and `mstats` I decided to only declare the `namedtuple` in `stats` and use for both. This for code reuse and consistency.
2) I aim to make sure the tests cover referencing by index and explicitly by attribute name as well as testing for the name of the `namedtuple`.
3) I wasn't sure on whether to keep or drop the leading underscore in the `namedtuple` declaration (e.g. `_DescribeResult`). I think it might be a source of confusion since it is being reused elsewhere but it also helps avoid name collisions just in case. I would like to hear opinions on this.
4) My aim is to address all the listed functions in #4192. This includes functions in `stats`, `mstats_basic`, `morestats` and possibly `_binned_statistic` although I'm not sure about `_multivariate` (there is an occurrence of `mode` there that I need to have a look at).

clarkfitzg commented:
Thanks for following up on this.

+1 for leading underscore- it keeps things out of the namespace.

clarkfitzg commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -420,6 +420,17 @@ def test_mode(self):
+
+        # test namedtuple attributes explicitly by name
+        assert_equal(a1_res.mode, getattr(a1_res, 'mode'))
+        assert_equal(a1_res.count, getattr(a1_res, 'count'))
```
It's not clear to me what this is testing? Does it do anything beyond what the 3 lines above it do?

Rationale- if `getattr(a1_res, 'mode')` doesn't raise `AttributeError` then it should always be true that `a1_res.mode == getattr(a1_res, 'mode'))`

Also, this could be in it's own test function.

aeklant replied:
Yes. This test is redundant; I wanted to make the `a1_res.mode` call explicit (in addition to `a1_res[0]`) but if this creates more confusion than clarity it should be scrapped.

I didn't do it in a different function for consistency within this module. I agree that there should be a class and the tests should be separated by functions but that would require rewriting the rest of the test module if we wanted to keep things consistent; I think that is out of scope for this PR (note that in the `test_stats` module, the tests are indeed done as you mentioned).

clarkfitzg commented:
When I was thinking about the implementation the testing seemed like the hardest part. What exactly are the important features to test, and how to do it in a way that doesn't involve tons of duplicated code?

aeklant commented:
I agree @clarkfitzg. Testing is certainly the most time consuming in this case but I do want to give it the proper attention, which is why I appreciate the input from more experienced community members. Thanks for your feedback.

rgommers commented:
What I would do is write a generic test function for this:

    def check_named_results(res, attr):
        """For function that return namedtuples, check return value names and ordering"""
        for i, attr in enumerate(attributes):
            assert_equal((res[i], getattr(res, attr))

and then for each function add a test that's as short as possible:

    def test_result_attr_names(self):
        res = stats.mode(np.arange(10))
        check_named_results(res, ['mode', 'count'])

rgommers commented:
> 1) For the functions that are in both stats and mstats I decided to only declare the namedtuple in stats and use for both. This for code reuse and consistency.

Makes sense.

> 2) I aim to make sure the tests cover referencing by index and explicitly by attribute name as well as testing for the name of the namedtuple.

The name of the namedtuple itself is only used for repr/str representation, so doesn't need to be tested I think.

> 3) I wasn't sure on whether to keep or drop the leading underscore in the namedtuple declaration (e.g. _DescribeResult). I think it might be a source of confusion since it is being reused elsewhere but it also helps avoid name collisions just in case. I would like to hear opinions on this.

Also +1 for leading underscore

> 4) My aim is to address all the listed functions in #4192. This includes functions in stats, mstats_basic, morestats and possibly _binned_statistic although I'm not sure about _multivariate (there is an occurrence of mode there that I need to have a look at).

``_multivariate`` contains only distributions, so no need to look into those.

aeklant commented:
@rgommers I think the generic test is a great idea. Is there maybe a centralised place to define the function only once (instead of once for every test module that may require it). `common_tests` perhaps? I notice `stats` is imported by most of the testing modules but that's hardly a good place for this.

rgommers commented:
``common_tests`` should be fine. There's only functions for distributions in there now, but no good reason not to add this function there I think.

ev-br commented:
There seems to be too little bikeshedding going on here :-). I'm not sure about the leading underscore. `_DescribeResult` is something a user directly interacts with, so it's not private irrespective of how we define "private". 

 OTOH, it's a bit weird to add over two dozens "public" names (where "public" means documented and guaranteed to be stable) just for holding return values from all these functions. 

Maybe the right thing to do is to document that all these functions return a namedtuple `DummyStatsResult` and the specific attributes of which depend on the function. Then we only add one name to the namespace, generate the required namedtuples *inside* the functions, and document the fields (which we do anyway at the moment).

(ducks and hides)

clarkfitzg commented:
+1 for a general test function.

@ev-br, are you suggesting something like this?
```
from collections import namedtuple

def mode(a, axis=0):
    '''
    Just for illustration
    '''
    ModeResult = namedtuple('ModeResult', ('mode', 'count'))
    return ModeResult(10, 20) 

b = mode(10)
print(str(b), repr(b), type(b))
```
Which generates:
```
('ModeResult(mode=10, count=20)', 'ModeResult(mode=10, count=20)', <class '__main__.ModeResult'>)
```

This leaves the namespace clean. If a user starts calling `mode` the return objects are of class `ModeResult`. Simple enough. This naming convention could then even be used for every function.

aeklant commented:
@ev-br: This is why I thought it may be a source of confusion and dropping the underscore may be a good enough idea to discuss. Also +1 for bikeshedding. I never felt so much a part of the community (tears of joy).

I agree that moving the namedtuples definition to *inside* the functions (as exemplified by @clarkfitzg) has its merits but I see two drawbacks which I think are worth discussing:

1) In this sample I am reusing the `stats._ModeResult` in `mstats` and that will no longer be possible if we move it to inside the function. This means duplicated code.
2) As a result of the duplicated code, the maintenance effort will also increase. I am referring specifically to the topic of having `stats` functions return Bunch objects instead of namedtuples in the future (see issue #3665).

Now, I don't mind doing a little extra work in this PR due to  [1] but I think the maintenance effort from [2] should be considered, because we don't know who will implement the Bunch objects functionality or when, nor how the transition will be performed or its timeframe. 

If Bunches will be implemented in the short term, it may be worthwhile to keep the namedtuples outside of the functions to make the transition smoother for whoever will develop it but if its a longterm effort then inside the functions is probably better, to keep things tidy.

Thoughts?

clarkfitzg commented:
`namedtuple` preserves backwards compatibility. Not sure how `bunch` could be backwards compatible. Also I don't see the advantage to a `bunch` over just returning a dict.

@aeklant it might actually be *less* maintenance to have them in the function like this. If I'm a maintainer reading code and I see something like
```
return _ModeResult(x, y) 
```
where `_ModeResult` was defined somewhere else, or imported from some other module, then I would probably think that it has more functionality. If it's in the function, immediately above the return statement then the purpose and use would be clear to me. 

Other opinions?

rgommers commented:
I already typed up a review comment previously about putting the ``namedtuple`` call right above the return statements, but then I saw that https://docs.python.org/3/library/collections.html#collections.namedtuple always puts them outside so I thought "never mind". It is much easier to read though, and that names stay consistent between ``stats`` and ``mstats`` can be easily taken care of in the tests. So in principle a good idea.

rgommers commented:
@aeklant I don't think we're going to be able to go to bunches for existing code - the backwards compat break is too large.

aeklant commented:
Hello all,

I have made the following changes:

1) The `namedtuple` definition has been moved (by popular demand) to within the function that uses said `namedtuple` (and the leading underscore has been dropped). So a bit less code reuse but a fair amount of more clarity.
2) A generic test for `namedtuple` attributes has been added at `common_tests` for code reuse.

*Note:* 
Using the generic test for `mstats.describe` in `test_mstats_basic` produces the following error:
&nbsp;&nbsp;`File "[...]/lib/python2.7/site-packages/numpy/testing/utils.py", line 275, in assert_equal`
&nbsp;&nbsp;&nbsp;&nbsp;`return assert_array_equal(actual, desired, err_msg, verbose)`
&nbsp;&nbsp;`File "[...]/lib/python2.7/site-packages/numpy/testing/utils.py", line 739, in assert_array_equal`
&nbsp;&nbsp;&nbsp;&nbsp;`verbose=verbose, header='Arrays are not equal')`
&nbsp;&nbsp;`File "[...]/lib/python2.7/site-packages/numpy/testing/utils.py", line 646, in assert_array_compare`
&nbsp;&nbsp;&nbsp;&nbsp;`val = comparison(x, y)`
&nbsp;&nbsp;`File "[...]/lib/python2.7/site-packages/numpy/ma/core.py", line 3707, in __eq__`
&nbsp;&nbsp;&nbsp;&nbsp;`check._mask = mask`
`AttributeError: 'numpy.bool_' object has no attribute '_mask'`

This doesn't happen if I do the `assert_equal` directly inside the test instead of calling the generic function so I have left it like that for now. I think it may have to do with the inner workings of `getattr` (which I will investigate a little further, since I don't remember all the nuances at the moment) or perhaps with the inner workings of `assert_equal` in `np.testing`. If you have a clue, please do let me know.

PS. Sorry for the delayed response. I found myself getting a little jittery with the GSoC decision approaching and decided to take a step back until I could get to business a bit more relaxed. I am getting back on the horse now!

rgommers commented:
The exception you saw is due to ``np.testing.assert_equal`` not being able to deal with masked arrays; there's ``np.ma.testutils.assert_equal`` instead: https://github.com/numpy/numpy/blob/master/numpy/ma/testutils.py#L82

So when you get to adding more namedtuples to ``mstats`` functions (not soon I think?) you can create a second common check function. For now it's OK as is I think.

rgommers commented:
This PR is now in good shape and could be merged, but I'm inclined to do so only after branching 0.16.x. It would be better to convert all the functions in gh-4192 at once.

@aeklant do you plan to put all functions in this PR, or do you want to work in a second PR for that?

aeklant commented:
@rgommers, I was thinking of doing all the functions in the same PR. This includes `stats`, `mstats`, `morestats` and possibly `_binned_statistic`. It is also a milestone in my GSoC roadmap so unless someone has an objection, I wouldn't mind getting it done before the coding period officially starts.

About `np.ma.testutils.assert_equal`, how about adding a boolean parameter like `ma=False` to the generic function? That way we can keep it in one function and just add `ma=True` for masked arrays instead of having two functions that do basically the same thing with a minor difference. Thoughts?

clarkfitzg commented:
Thanks!

On Thu, Apr 30, 2015 at 2:13 PM, Abraham Escalante <notifications@github.com
> wrote:

> @rgommers <https://github.com/rgommers>, I was thinking of doing all the
> functions in the same PR. This includes stats, mstats, morestats and
> possibly _binned_statistic. It is also a milestone in my GSoC roadmap so
> unless someone has an objection, I wouldn't mind getting it done before the
> coding period officially starts.
>
> About np.ma.testutils.assert_equal, how about adding a boolean parameter
> like ma=False to the generic function? That way we can keep it in one
> function and just add ma=True for masked arrays instead of having two
> functions that do basically the same thing with a minor difference.
> Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy/scipy/pull/4709#issuecomment-97968381>.
>

aeklant commented:
I made the change to handle masked arrays in the same function. Let me know if you guys find this to be a good approach or if you think I should make two distinct functions instead. 

If there are no more corrections or adjustments I think all that's left to do would be to rinse & repeat for the rest of the tests and maybe document the `check_named_results` function.

rgommers commented:
The change to ``check_named_results`` looks good to me.

rgommers commented:
Makes sense to put it all in this PR. I agree that it's now mostly just more of the same.

rgommers commented:
I've edited the title of this PR to start with ``WIP``, to make it clear to everyone that this is not to be merged yet. Once you've added all functions, you can change it back to ``ENH``.

aeklant commented:
I have made the changes to `stats.py`. I would appreciate a fresh set of eyes making a review, especially since I updated the documentation to reflect the `namedtuple` names. Just to make sure I didn't misunderstand any of the names that were agreed upon or missed a typo.

The tests ran fine so I will continue with `mstats`. The merge conflicts probably have to do with the branch going too far behind the main repository's master, right? Should I try to pull the latest master to the branch or maybe wait until the work is done?

*Note:*
`stats.ranksums` did not have any test coverage so I added it but only the `namedtuple` test case. I should add more when I deal with #638.

rgommers commented:
The merge conflict is indeed because another PR was merged into master. It's easy to resolve. Don't merge master into your feature branch though, that gets messy over time. Usually (and here as well) you just want to rebase onto latest master and resolve the merge conflict. So:

    $ git fetch upstream
    $ git rebase upstream/master  # stops at a merge conflict, edit stats.py and test_stats.py
    $ git add scipy/stats/stats.py
    $ git add scipy/stats/tests/test_stats.py
    $ git rebase --continue  # all OK now (but do check)
    $ git push origin issue_#4192 -f  # need a force push to github after any rebase

rgommers commented:
Changed in the last commit look good to me from a quick browse.

aeklant commented:
Alright, I think I'm pretty much done with the changes. They are pretty straight forward after the consensus but they turned out to be very time consuming too so again, I would appreciate a peer review.

*A few notes:*
1) `mstats.f_oneway`, `mstats.mannwhitneyu` and `mstats.kruskal` had no test coverage, much like `stats.ranksums`, so I added a test for the `namedtuple` attributes but more extensive coverage is outside the scope of this PR so I will leave that for when I deal with the specific **StatisticsCleanup** issues for each of those functions.

2) `morestats.shapiro` does not follow the same pattern (as pointed out by @clarkfitzg in gh-4192). Sometimes it returns two values, sometimes three depending on the case.
I was thinking of padding the two value result with a default `None` at the end for the third value: `ShapiroResult(w, pw, None)`.
But this would probably break a lot of already written code (anything in the form `x, y = shapiro(*args)`) so I don't know if we should keep the namedtuples out of this one or try a workaround.

3) `morestats.bayes_mvs` encompasses three results so I made a `namedtuple` for each but I didn't use another (more global) one (i.e. `Bayes_mvsResult`) because I found that `namedtuple`s within a `namedtuple` become a little hard to read, but let me know what you think. The output currently looks like this: `(mean(estimate=7.0, minmax=(4.9662142379679404, 9.0337857620320587)), variance(estimate=23.333333333333332, minmax=(11.821932327635709, 42.613865189698316)), std_dev(estimate=4.7309785331488898, minmax=(3.438303699156855, 6.5279296250571139)))`.
Also, this is the one implementation that I actually had to change quite a bit in comparison to the rest so a bit more attention here for the peer review should be in order.

aeklant commented:
Let me know if I should rebase and squash to make the commit history a little neater before merging. 

Cheers.

clarkfitzg commented on scipy/stats/morestats.py:
```diff
@@ -78,11 +79,20 @@ def bayes_mvs(data, alpha=0.90):
                          % alpha)
-    return tuple((x.mean(), x.interval(alpha)) for x in res)
+
+    mean = namedtuple('mean', ('estimate', 'minmax'))
```
Capitalize? Also, `statistic` might be more consistent with the rest of the changes name compared to `estimate`

rgommers commented on scipy/stats/_binned_statistic.py:
```diff
@@ -146,7 +146,9 @@ def binned_statistic(x, values, statistic='mean',
                                              bins, range)
 
-    return medians, edges[0], xy
+    Binned_statisticResult = namedtuple('Binned_statisticResult',
```
PEP8 frowns on this, maybe change to ``'BinnedStatisticResult'``

rgommers replied:
and same for ``2d`` and ``dd`` versions below

rgommers commented on scipy/stats/_binned_statistic.py:
```diff
@@ -204,9 +206,9 @@ def binned_statistic_2d(x, y, values, statistic='mean',
     statistic : (nx, ny) ndarray
         The values of the selected statistic in each two-dimensional bin
-    xedges : (nx + 1) ndarray
+    x_edge : (nx + 1) ndarray
```
I like the underscore, but these return values do contain multiple edges. so add the ``s`` back at the end?

rgommers commented on scipy/stats/mstats_basic.py:
```diff
@@ -1971,7 +2014,13 @@ def f_oneway(*args):
+    -------
+    statistic : float
+        The computed F-value of the test.
+
```
this blank line can be removed

rgommers commented on scipy/stats/mstats_basic.py:
```diff
@@ -2023,7 +2074,13 @@ def friedmanchisquare(*args):
+    -------
+    statistic : float
+        the test statistic.
+
```
this blank line can be removed

rgommers commented on scipy/stats/mstats_basic.py:
```diff
@@ -2023,7 +2074,13 @@ def friedmanchisquare(*args):
+        the test statistic.
+
+    pvalue : float
+        the ssociated p-value.
```
typo --> associated

rgommers commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2048,6 +2113,11 @@ def test_ks_2samp():
 
+    # test for namedtuple attribute results
+    attributes = ('statistic', 'pvalue')
+    res = stats.ks_2samp(data1-0.01, data2)
```
nitpick: spaces around ``-``

rgommers commented:
I went through all changes - overall looks good, can be merged once the last comments are addressed I think. No need to rebase/squash, 7 commits for a PR this size is not that much and your commit messages look fine.

rgommers commented:
Re: (1) I agree with leaving extending test coverage of now untested functions until later. This PR is large enough as is.

rgommers commented:
Regarding ``shapiro``, I guess you could return different namedtuples for the 2 vs 3 return values. But it becomes a bit ugly. Maybe just don't touch it.

rgommers commented:
What you made ``bayes_mvs`` return looks fine to me. A namedtuple of namedtuples would indeed count as going overboard in my book:)

rgommers commented:
@aeklant looks like we're almost there. Release split will probably be on Sunday; this could still make it in.

aeklant commented:
@rgommers last comments have been addressed. Do let me know if you find any other details in this or the previous commits.

rgommers commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -35,6 +36,16 @@
+class TestBayes_mvs(TestCase):
+    def test_result_attributes(self):
+        x = np.arange(15)
+        attributes = ('estimate', 'minmax')
```
``estimate`` should be ``statistic`` here

rgommers commented:
No idea why TravisCI turned green on the last commit, it should have detected this:

    ======================================================================
    ERROR: test_result_attributes (test_morestats.TestBayes_mvs)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/rgommers/Code/scipy/scipy/stats/tests/test_morestats.py", line 46, in test_result_attributes
        check_named_results(i, attributes)
      File "/home/rgommers/Code/scipy/scipy/stats/tests/common_tests.py", line 22, in check_named_results
        npt.assert_equal(res[i], getattr(res, attr))
    AttributeError: 'Mean' object has no attribute 'estimate'

All other tests pass for me.

aeklant commented:
I made the mistake of not running the tests before commiting because they we're so simple. It won't happen again. It should be fixed now.

rgommers commented:
@aeklant no worries, that's happened to all of us.

rgommers commented:
All right, in it goes. Thanks @aeklant, this is a nice improvement! And thanks @clarkfitzg and @josef-pkt for the inputs on this.

aeklant commented:
Great! Thanks @rgommers. And thanks to all who contributed.

clarkfitzg commented:
Nice work here @aeklant @rgommers. 

One question- how to maintain this consistency in naming going forward? Might be nice to be able to look up these standard names somewhere in the docs: https://docs.scipy.org/doc/scipy/reference/stats.html

rgommers commented:
@clarkfitzg you mean consistency when new functions get added? And you're suggesting a separate section in the refguide with "this is the pattern of naming return values"?

clarkfitzg commented:
@rgommers Yes, or something along those lines. Or even a preventative measure to avoid creating a new function that returns `pval` rather than `pvalue`. 

Maybe this is unnecessary?

rgommers commented:
I'd say for new functions we probably won't want ``namedtuple``s anyway, but dicts, bunches or custom classes. Something extendible at least. A test function to check for unwanted names is overkill I think - that's what code review is for.

bsipocz commented:
I'm very sorry to comment on this long merged PR. If you think this is a real issue, I'm happy to open one for it.

Today while updating some doctest I've run into these namedtuples, and wondered why they are not mentioned at all in the docstrings neither in the examples in the docstrings, so I needed to look into the source to see what's going on.

Is it an oversight or a decision?
Alternatively as I have similar examples that I don't necessarily want to update, I'm very much interested how you make the doctests pass for lines like e.g. https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/stats.py#L4358?

ev-br commented:
We customize the doctester, https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L587

Discussion is indeed better on a new issue (visibility)
